### ISSUE TITLE
The legal policy normalizes the whole action vocabulary to read the 1.6% of it that is legal

### ISSUE THREAD

**Issue by mmmfrieddough** [mmmfrieddough/anthro-chess#306]
## What was observed

Found while working #287, in the two lines that change left alone.

**`legal_policy_log_probabilities` builds two full-width float64 tensors to
read a fiftieth of one of them.** After #287 the gather is 0.50 ms and the
normalize in front of it is the rest:

```python
masked = active.logits.masked_fill(~active.legal_mask, -torch.inf)
normalized = torch.log_softmax(masked, dim=-1)
```

At the real evaluation batch shape — 610 enabled rows, vocabulary 1970 — those
two lines materialize two fresh 9.6 MB tensors over 1,201,700 entries in order
to read the 18,778 legal ones, **1.6% of what was computed**. Measured on this
host:

| variant | ms per call |
| --- | ---: |
| before #287 | 8.30 |
| #287 as merged | 3.02 |
| normalizing only the gathered entries | **0.41** |

Roughly 2.3 ms of the 2.87 ms is allocation rather than arithmetic: the call
takes 4,956 minor page faults against 2.3 for the segment form. Forcing the
allocator warm with `MALLOC_MMAP_THRESHOLD_=1G` drops the current form to
0.71 ms and leaves the segment form at 0.41 ms, so the win is 7.3x in the
default glibc regime and 1.7x at the floor.

It runs three times per batch, so this is **9.1 ms to 1.25 ms per batch**.

## Why

Same family as #211, #284 and #287, one level down: after the per-position loop
went, the remaining cost in this function is work done over the whole
vocabulary for an answer defined over the legal actions only.

## What it does *not* buy

**This one moves the recorded numbers.** A log-sum-exp by segment over the
gathered entries agrees with the full-width normalize to within 2.8e-14, and
about 87% of values come out bitwise identical — but the rest do not. This
module normalizes in float64 specifically because its readings are compared
across checkpoints and machines, so this is a deliberate trade rather than a
free win, and it is the reason this is a separate issue rather than part of
#287.

Whoever takes it should decide that question first: whether a shift at the
fourteenth decimal is acceptable in a quantity that feeds `policy_divergence`
and the dependency signals, or whether the allocation cost is the price of a
reading that reproduces exactly.

## What would settle it

Compute the per-row maximum and the log-sum-exp by segment over the legal
entries alone, with `scatter_reduce_` and `scatter_add_` over a row-index
vector. Note that #287 settled on gathering through `ActiveBatch.legal_mask`
rather than carrying an index, so that row vector has to be derived here —
`repeat_interleave` over the row lengths, about 54 us per batch at this shape —
or `ActiveBatch` has to start carrying it, which #287 measured and declined for
its own purposes but which this change and #305 would both use.

Then answer the reproducibility question above with a comparison against the
current values on a real view, and record which way it went.

## Measurement context

One Linux host, 2x RTX 4090, 32 cores, CUDA, both GPUs idle. CPU float64, at
tensor shapes taken from real batches over `lichess-blitz-2017-04-test`.

## References

- #287 — removed the per-position gather in the same function and left this
- #305 — the other half of what #287 left, in the consumer; wants the same row
  vector this one does

**Comment by mmmfrieddough** [comment]
Claimed for implementation by Claude Code.

Session: 1693c363-8ed0-415b-9c63-b12403a1f878

**Comment by mmmfrieddough** [comment]
Correction to this issue's timing table, found while verifying the change.

**The 7.3x figure is a fresh-process measurement; the real evaluation process runs at the 1.7x floor this issue already names.** In situ the change is worth about 19%, not 5-7x.

Timed inside three real `anthro eval run` invocations per arm, interleaved, summing every call over a 128-game view of `lichess-blitz-2017-04-test` (57 calls, median 410 enabled rows):

| arm | time in `legal_policy_log_probabilities` per view | per call |
| --- | ---: | ---: |
| before | 161.4 / 152.0 / 151.5 ms | 2.77 ms |
| after (segment normalize) | 130.3 / 123.5 / 123.9 ms | 2.25 ms |

19% off the function, 1.23x. That is **2.4% of the dependency pass** and **0.11% of the command**, and at pass level it does not clear noise — the trajectory pass reads 1.188 s after against 1.190 s before.

The mechanism is the one this issue identified, just with the sign of the conclusion changed by which regime is real. Verified directly: the *same* captured `ActiveBatch`, same code, same 16 threads, costs **2.16 ms per call in a clean harness and 0.52 ms inside the eval process**. The evaluation process holds a large working heap from the model and the loader, so glibc satisfies the 6.5 MB `masked_fill` and `log_softmax` allocations out of the existing arena instead of via mmap/munmap. The page-fault cost that dominates an isolated benchmark is mostly not paid where the code actually runs.

This does not change what #307 measured for #287 — that change removed per-position Python work, which is allocator-independent — but it does mean **a micro-benchmark of an allocation-heavy tensor path in a fresh process will overstate the win here**, and the same caution applies to the figures in #305's body.

The change still stands on the other half of this issue's question: the segment form is bitwise reproducible across vector widths where the full-width form is not. That is written up in #315.

Measured on one Linux host, 2x RTX 4090 both idle, 32 cores, CPU float64.

**Comment by mmmfrieddough** [comment]
Closing as not planned. The question this issue posed was answered; the answer is that the change is not worth making.

This issue framed the trade as speed against a last-bit shift, and asked whoever took it to decide the reproducibility question first. Both halves came apart under measurement.

**Speed — 0.11% of the command, not 7x.** In situ, across three interleaved runs per arm over a 128-game view, the normalize goes from 155 ms to 126 ms: 19% of the function, 2.4% of the dependency pass, and a tie at pass level (1.188 s against 1.190 s). This issue already named the mechanism — "7.3x in the default glibc regime and 1.7x at the floor" — and the real evaluation process runs at the floor, because it holds a large working heap from the model and loader. The same captured `ActiveBatch`, same code, same threads: **2.16 ms per call in a clean harness, 0.52 ms inside the eval process.**

**Reproducibility — the premise was wrong in both directions.** I first found that the segment form is bitwise stable across vector widths where the full-width form is not (7 dependency fields move between an AVX2 and a scalar dispatch, worst 2.1e-15) and argued that made the change *more* reproducible. That argument does not hold:

- `ATEN_CPU_CAPABILITY` is a runtime dispatch knob on a single binary, not a build. Forcing it low is artificial, and no machine that would run this lacks AVX2.
- The forward pass runs on **CUDA in float32**. Logits reach this function with ~1e-7 relative precision and vary with GPU architecture, library version and batch shape — `docs/evaluation.md:1898` records ~3e-6 movement from batching alone. Stabilizing the fifteenth decimal of a value whose inputs are good to the seventh buys nothing. Cross-machine bitwise reproducibility of these readings does not exist and this would not have created it.

So the cost — replacing `log_softmax(masked_fill(...))` with a manual max-shift, `scatter_reduce_` and `scatter_add_` over a derived segment vector, plus a helper and an overflow-guard test — buys 0.11% and a numerical property with no consumer.

Implementation and full measurements are in #315, closed unmerged; the branch remains on origin.

**Two findings worth carrying forward:**

- **#318** — filed from this issue's shakedown reading, and unrelated to it: `_maturity` takes `step` from the checkpoint but `processed_positions` from the run record, so every checkpoint but a run's last reports the run's final count. 8,757,080 against a true 1,092,677 at step 1000, in the coordinate dependency results are read against.
- **Time this family of work in situ.** A fresh-process benchmark of an allocation-heavy tensor path overstates the win by roughly 4x here. The caution is recorded on #305, whose own 13.9 ms figure is a fresh-process number and should be re-measured before that work is sized.
